pkg/ottl/contexts/internal/ctxcontext/context.go

```diff
@@ -0,0 +1,306 @@
+	}
+}
+
+func convertAuthDataToMap(authData client.AuthData) (pcommon.Map, error) {
+	authMap := pcommon.NewMap()
+	if authData == nil {
+		return authMap, nil
+	}
+	names := authData.GetAttributeNames()
+	for _, name := range names {
+		attrVal := authData.GetAttribute(name)
+		attrStr, err := getAuthAttributeValue(attrVal)
+		if err != nil {
+			return pcommon.NewMap(), err
+		}
+		switch v := attrStr.(type) {
+		case string:
+			authMap.PutStr(name, v)
+		default:
+			authMap.PutEmpty(name)
+		}
+	}
+	return authMap, nil
+}
+
+func accessClientAuthAttributesKeys[K any]() ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, _ K) (any, error) {
+			cl := client.FromContext(ctx)
+			return convertAuthDataToMap(cl.Auth)
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.auth.attributes")
+		},
+	}
+}
+
+func accessClientAuthAttributesKey[K any](keys []ottl.Key[K]) ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, tCtx K) (any, error) {
+			if len(keys) == 0 {
+				return nil, errors.New("cannot get map value without keys")
+			}
+			cl := client.FromContext(ctx)
+			key, err := ctxutil.GetMapKeyName(ctx, tCtx, keys[0])
+			if err != nil {
+				return nil, err
+			}
+			if cl.Auth == nil {
+				return nil, nil
+			}
+			attrVal := cl.Auth.GetAttribute(*key)
+			attrStr, err := getAuthAttributeValue(attrVal)
+			if err != nil {
+				return nil, err
+			}
+			return attrStr, nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.auth.attributes")
+		},
+	}
+}
+
+func convertClientMetadataToMap(md client.Metadata) pcommon.Map {
+	mdMap := pcommon.NewMap()
+	for k := range md.Keys() {
+		sl := mdMap.PutEmptySlice(k)
+		mdVal := md.Get(k)
+		mdSl := convertStringArrToValueSlice(mdVal)
+		mdSl.Slice().CopyTo(sl)
+	}
+	return mdMap
+}
+
+func accessClientMetadataKeys[K any]() ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, _ K) (any, error) {
+			cl := client.FromContext(ctx)
+			return convertClientMetadataToMap(cl.Metadata), nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.metadata")
+		},
+	}
+}
+
+func accessClientMetadataKey[K any](keys []ottl.Key[K]) ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, tCtx K) (any, error) {
+			if len(keys) == 0 {
+				return nil, errors.New("cannot get map value without keys")
+			}
+
+			key, err := ctxutil.GetMapKeyName(ctx, tCtx, keys[0])
+			if err != nil {
+				return nil, fmt.Errorf("cannot get map value: %w", err)
+			}
+			cl := client.FromContext(ctx)
+			mdVal := cl.Metadata.Get(*key)
+			if len(mdVal) == 0 {
+				return nil, nil
+			}
+			return getIndexableValueFromStringArr(ctx, tCtx, keys, mdVal)
```

Comment:
```suggestion
			return getIndexableValueFromStringArr(ctx, tCtx, keys[1:], mdVal)
```

Same as my previous comment.